Library/Homebrew/os.rb

```diff
@@ -1,6 +1,6 @@
 module OS
   def self.mac?
-    /darwin/i === RUBY_PLATFORM
+    /darwin/i === RUBY_PLATFORM || ENV["HOMEBREW_EMULATE_MAC"]
```

Comment:
I think what you are saying is a goal Homebrew could and should strive for. But I think this still requires a lot of work, as not everything in a formula is purely declarative. One example I can think of is dependencies being declared (or not) based on the currently running OS X release by using a suffixed `if` clause. There are probably many other things that require thought and redesign …